File: cpp/src/verify.cpp
/// Verify: run 4str_0int ht=65536 lf=0.5 sel=0.1, print NumGroups.
#include <cstdio>
#include <cstdint>
#include <vector>
#include <string>
#include <numeric>
#include <random>
#define XXH_INLINE_ALL
#include "xxhash.h"
#include "column_marshaller.h"

static inline uint64_t HB(const uint8_t* d, size_t l, uint64_t s) { return XXH3_64bits_withSeed(d,l,s); }

int main() {
    constexpr size_t NS = 4, HT = 65536, NP = 1000000;
    constexpr double LF = 0.5, SEL = 0.1;
    size_t NK = static_cast<size_t>(HT * LF);

    std::mt19937_64 rng(42);

    // Build keys
    std::vector<std::vector<std::vector<uint8_t>>> strCols(NS);
    for (size_t c = 0; c < NS; c++) {
        strCols[c].resize(NK);
        for (size_t i = 0; i < NK; i++) {
            auto s = "key_" + std::to_string(i) + "_c" + std::to_string(c);
            strCols[c][i].assign(s.begin(), s.end());
        }
    }
    std::vector<int64_t> buildHashes(NK);
    for (size_t i = 0; i < NK; i++) {
        uint64_t h = 0;
        for (size_t c = 0; c < NS; c++) h = HB(strCols[c][i].data(), strCols[c][i].size(), h);
        buildHashes[i] = static_cast<int64_t>(h);
    }
    std::vector<int64_t> buildValues(NK);
    for (size_t i = 0; i < NK; i++) buildValues[i] = i % 1000;

    // Probe keys
    size_t nH = static_cast<size_t>(NP * SEL), nM = NP - nH;
    std::vector<std::vector<std::vector<uint8_t>>> probeStr(NS);
    for (auto& v : probeStr) v.reserve(NP);
    std::vector<int64_t> probeHashes; probeHashes.reserve(NP);

    for (size_t i = 0; i < nH; i++) {
        size_t idx = rng() % NK;
        for (size_t c = 0; c < NS; c++) probeStr[c].push_back(strCols[c][idx]);
        probeHashes.push_back(buildHashes[idx]);
    }
    for (size_t i = 0; i < nM; i++) {
        uint64_t h = 0;
        for (size_t c = 0; c < NS; c++) {
            auto s = "miss_" + std::to_string(i) + "_" + std::to_string(c);
            std::vector<uint8_t> b(s.begin(), s.end());
            h = HB(b.data(), b.size(), h);
            probeStr[c].push_back(std::move(b));
        }
        probeHashes.push_back(static_cast<int64_t>(h));
    }

    // Shuffle
    std::vector<size_t> ord(NP); std::iota(ord.begin(), ord.end(), 0);
    for (size_t i = NP - 1; i > 0; i--) std::swap(ord[i], ord[rng() % (i + 1)]);
    std::vector<std::vector<std::vector<uint8_t>>> shuffledStr(NS);
    for (size_t c = 0; c < NS; c++) { shuffledStr[c].resize(NP); for (size_t i = 0; i < NP; i++) shuffledStr[c][i] = std::move(probeStr[c][ord[i]]); }
    std::vector<int64_t> shuffledHashes(NP); for (size_t i = 0; i < NP; i++) shuffledHashes[i] = probeHashes[ord[i]];
    std::vector<int64_t> probeValues(NP); for (size_t i = 0; i < NP; i++) probeValues[i] = i % 1000;

    // Combine
    size_t TR = NK + NP;
    for (size_t c = 0; c < NS; c++) for (auto& v : shuffledStr[c]) strCols[c].push_back(std::move(v));
    std::vector<int64_t> allHashes = buildHashes; allHashes.insert(allHashes.end(), shuffledHashes.begin(), shuffledHashes.end());
    std::vector<int64_t> allValues = buildValues; allValues.insert(allValues.end(), probeValues.begin(), probeValues.end());

    // Build VarcharSlice arrays
    std::vector<std::vector<taper::VarcharSlice>> slices(NS);
    for (size_t c = 0; c < NS; c++) {
        slices[c].resize(TR);
        for (size_t i = 0; i < TR; i++) { slices[c][i].ptr = strCols[c][i].data(); slices[c][i].len = strCols[c][i].size(); }
    }

    // Run taper
    size_t distinctKeys = NK + nM;
    size_t minSlots = std::max(static_cast<size_t>(distinctKeys / 0.85), size_t(8));
    size_t numChunks = 1; while (numChunks * 8 < minSlots) numChunks *= 2;

    taper::SimpleArenaAllocator pool;
    std::vector<taper::ColumnDesc> cd(NS, taper::ColumnDesc::Varchar);
    taper::TaperColumnSerializeHandler t(pool, 8, cd, numChunks);

    constexpr size_t BATCH_SIZE = 410;
    size_t numBatches = (TR + BATCH_SIZE - 1) / BATCH_SIZE;
    std::vector<taper::ColumnInput> cols(NS);

    for (size_t batch = 0; batch < numBatches; batch++) {
        size_t start = batch * BATCH_SIZE;
        size_t end = std::min(start + BATCH_SIZE, TR);
        int32_t batchLen = static_cast<int32_t>(end - start);
        for (size_t c = 0; c < NS; c++) cols[c] = taper::ColumnInput::MakeVarchar(slices[c].data() + start);
        t.EmplaceTableWithDecode(allHashes.data() + start, batchLen, cols, allValues.data() + start);
    }

    printf("C++:  NumGroups = %zu\n", t.NumGroups());
}

File: rust/examples/verify.rs
/// Verify: run 4str_0int ht=65536 lf=0.5 sel=0.1, print NumGroups + agg checksum.
use taper_hashmap::column_marshaller::{TaperColumnSerializeHandler, ColumnDesc, ColumnInput};
use rand_mt::Mt19937GenRand64;
use xxhash_rust::xxh3::xxh3_64_with_seed;

fn hash_bytes(data: &[u8], seed: u64) -> u64 { xxh3_64_with_seed(data, seed) }

fn main() {
    let num_str_cols = 4usize;
    let ht_size = 65536usize;
    let load_factor = 0.5f64;
    let selectivity = 0.1f64;
    let num_probe_rows = 1_000_000usize;
    let num_keys = (ht_size as f64 * load_factor) as usize;

    let mut rng = Mt19937GenRand64::new(42);

    // Build keys
    let build_str_cols: Vec<Vec<Vec<u8>>> = (0..num_str_cols)
        .map(|c| (0..num_keys).map(|i| format!("key_{}_c{}", i, c).into_bytes()).collect())
        .collect();
    let build_hashes: Vec<u64> = (0..num_keys)
        .map(|i| {
            let mut h = 0u64;
            for c in 0..num_str_cols { h = hash_bytes(&build_str_cols[c][i], h); }
            h
        }).collect();
    let build_values: Vec<i64> = (0..num_keys).map(|i| (i % 1000) as i64).collect();

    // Probe keys
    let num_hits = (num_probe_rows as f64 * selectivity) as usize;
    let num_misses = num_probe_rows - num_hits;

    let mut probe_str_cols: Vec<Vec<Vec<u8>>> = vec![Vec::with_capacity(num_probe_rows); num_str_cols];
    let mut probe_hashes: Vec<u64> = Vec::with_capacity(num_probe_rows);

    for _ in 0..num_hits {
        let idx = (rng.next_u64() as usize) % num_keys;
        for c in 0..num_str_cols { probe_str_cols[c].push(build_str_cols[c][idx].clone()); }
        probe_hashes.push(build_hashes[idx]);
    }
    for i in 0..num_misses {
        let mut h = 0u64;
        for c in 0..num_str_cols {
            let s = format!("miss_{}_{}", i, c).into_bytes();
            h = hash_bytes(&s, h);
            probe_str_cols[c].push(s);
        }
        probe_hashes.push(h);
    }

    // Shuffle
    let mut order: Vec<usize> = (0..num_probe_rows).collect();
    for i in (1..num_probe_rows).rev() { order.swap(i, (rng.next_u64() as usize) % (i + 1)); }
    let probe_str_cols: Vec<Vec<Vec<u8>>> = (0..num_str_cols)
        .map(|c| order.iter().map(|&i| probe_str_cols[c][i].clone()).collect()).collect();
    let probe_hashes: Vec<u64> = order.iter().map(|&i| probe_hashes[i]).collect();
    let probe_values: Vec<i64> = (0..num_probe_rows).map(|i| (i % 1000) as i64).collect();

    // Combine
    let mut all_str_cols = build_str_cols;
    for c in 0..num_str_cols { all_str_cols[c].extend(probe_str_cols[c].iter().cloned()); }
    let mut all_hashes: Vec<u64> = build_hashes; all_hashes.extend_from_slice(&probe_hashes);
    let mut all_values: Vec<i64> = build_values; all_values.extend_from_slice(&probe_values);

    let total_rows = all_hashes.len();

    // Run taper
    let num_misses_actual = num_probe_rows - num_hits;
    let distinct_keys = num_keys + num_misses_actual;
    let min_slots = (distinct_keys as f64 / 0.85) as usize;
    let num_chunks = ((min_slots + 7) / 8).next_power_of_two();

    let mut col_descs: Vec<ColumnDesc> = Vec::new();
    for _ in 0..num_str_cols { col_descs.push(ColumnDesc::Varchar); }

    let mut table = TaperColumnSerializeHandler::new(&col_descs, 8, num_chunks);

    const BATCH_SIZE: usize = 410;
    let num_batches = (total_rows + BATCH_SIZE - 1) / BATCH_SIZE;

    for batch_idx in 0..num_batches {
        let start = batch_idx * BATCH_SIZE;
        let end = (start + BATCH_SIZE).min(total_rows);
        let batch_hashes = &all_hashes[start..end];
        let batch_values = &all_values[start..end];
        let str_slices: Vec<Vec<&[u8]>> = (0..num_str_cols)
            .map(|c| all_str_cols[c][start..end].iter().map(|s| s.as_slice()).collect())
            .collect();
        let mut columns: Vec<ColumnInput> = Vec::new();
        for c in 0..num_str_cols { columns.push(ColumnInput::Varchar(&str_slices[c])); }
        table.emplace_table_with_decode(batch_hashes, &columns, batch_values);
    }

    println!("Rust: NumGroups = {}, AggChecksum = {}", table.num_groups(), table.agg_checksum());
}
